rust/src/treefile.rs

```diff
@@ -1153,6 +1187,75 @@ pub(crate) struct LegacyTreeComposeConfigFields {
     pub(crate) automatic_version_prefix: Option<String>,
 }
 
+#[derive(Serialize, Deserialize, Debug, Default, PartialEq, Eq)]
+#[serde(rename_all = "kebab-case")]
+pub(crate) struct DeriveCustom {
+    pub(crate) url: String,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub(crate) description: Option<String>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Default, PartialEq, Eq)]
+#[serde(rename_all = "kebab-case")]
+pub(crate) struct DeriveInitramfs {
+    pub(crate) regenerate: bool,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub(crate) etc: Option<Vec<String>>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub(crate) args: Option<Vec<String>>,
+}
+
+/// These fields are only useful when deriving from a prior ostree commit;
+/// at the moment we only use them when translating an origin file
+/// to a treefile for client side assembly.
+#[derive(Serialize, Deserialize, Debug, Default, PartialEq, Eq)]
+#[serde(rename_all = "kebab-case")]
+pub(crate) struct DeriveConfigFields {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub(crate) from: Option<String>,
```

Comment:
Though, thinking about this a bit more perhaps it should be `baseref` since we're actually writing an ostree ref to it, and we reserve `from` for container images?  Or, to disallow using both simultaneously, we make it a tagged enum?

Reply:
Or should we just stick with `baseref` for everything and encode the actual base layer type as part of the refspec? Similar to what we had with `rojig://`. Except for container images, it'd be... `oci://` I guess? That way, the field name itself is always the same, and it also matches what actually shows up in `rpm-ostree status` too.

So in theory `rpm-ostree rebase <refspec>` would just be about substituting the `baseref` and feeding it to the upgrader, without necessarily needing to parse the refspec.